File: 1807-evaluate-the-bracket-pairs-of-a-string/1807-evaluate-the-bracket-pairs-of-a-string.cpp
class Solution {
public:
    string evaluate(string s, vector<vector<string>>& knowledge) {
        map<string,string> ump;
        for(auto x:knowledge)
        {
            auto p=x[0],q=x[1];
            ump[p]=q;
        }
        string res="";
        string t="";
        bool found=false;
        for(int i=0;i<s.length();i++)
        {
            if(s[i]=='(')
            {
                found=true;
            }
            else if(s[i]==')')
            {
                found=false;
                if(ump.find(t)==ump.end())
                {
                    res+='?';
                }
                else
                {
                    res+=ump[t];
                }
                t="";
            }
            else
            {
                if(found) t+=s[i];
                else res+=s[i];
            }
        }
        return res;
    }
};

File: 1807-evaluate-the-bracket-pairs-of-a-string/1807-evaluate-the-bracket-pairs-of-a-string.py
class Solution:
    def evaluate(self, s: str, knowledge: list[list[str]]) -> str:
        dct = {}
        for ele in knowledge:
            dct[ele[0]] = ele[1]
        total_string = ""
        curr = ""
        for c in s:
            if c=='(':
                total_string+=curr
                curr=""
            elif c==')':
                total_string+=dct.get(curr,'?')
                curr=""
            else:
                curr+=c
        total_string+=curr
        return total_string
